### DISCUSSION TITLE
Can AVM2 classes contain nested classes?

### DISCUSSION THREAD

**Discussion by hydroper** [airsdk/Adobe-Runtime-Support#2524] [General]
I was reading the [AVM2 overview](https://jmendeth.com/snapshot/4d9475cfb10af8142e331551dc9b91e1217dc8c6/media/2014-05-17-reverse-engineering-flash/avm2overview.pdf) and wondered if it's allowed for a specific class to contain another class as trait.

Would AVM2 throw a `VerifyError` if it saw any nested class?

**Comment by ajwfrost** [comment]
That is a fascinating question...

So a quick try with the AS3 compiler shows "Classes must not be nested", which is thrown by the compiler when a class definition is found outside of the "global" scope. So, if the compiler is currently parsing a class definition, it cannot encounter a new class definition.

But that's just the compiler.. from an AS3 specification perspective (i.e. that document) it wasn't completely clear. But, I think there is a restriction that we come across during the creation of class definitions which would prevent this. Although it's possible to have the class/instance definitions which could have members that are classes, it's not possible to create these with ActionScript byte code, because a new class definition is created using the `newclass` instruction, and that says:

> When this instruction is executed, the scope stack must contain all the scopes of all base classes, as the scope stack is saved by the created ClassClosure. 

What that means is that when a new class is created, the scope stack is saved which then enables it to work out inheritance at a later point in time (e.g. if your class `B` is derived from `A` and you reference a property `myProp` that's defined in `A`, the scope stack is searched to find the storage for that property). But that's then incompatible with the idea of a new class definition appearing when you're also within the scope of an existing class..  Say you wanted your class `B` to contain a new class `C`; essentially this would mean that the scope stack for `C` would already contain `A` and `B` (plus whatever `C` derives from would have to have been pushed). And then, if within `C` you tried to access a property `myProp`, it would be found in the scope of `A` still... i.e. `C` would automatically derive from `B` and `A`....

In theory of course, that would all be possible i.e. if you took out the compiler restriction, you could perhaps create this sort of situation. In practice, you could then run into issues if your constructors didn't match but I'm not sure whether avmplus would inherently prevent such a thing...

But if you're talking about the generic possibility of nested classes like you can have anonymous/nested classes in Java, then this would more likely be handled via some workaround which makes you think you've got a nested class, without it actually being one (i.e. the compiler could generate a new class `B$C` within the global scope ?). I'm not sure whether that's feasible to do (would need a lot of changes to the compiler) or whether it's worthwhile (is there really a need for nested classes, vs the hierarchical package definitions?).

Interesting topic!
